File: 0187_repeated_dna_sequences/bitmask/python/solution.py
WINDOW_SIZE = 10
ENCODE = {"A": 0, "C": 1, "G": 2, "T": 3}


def find_repeated_dna_sequences(s: str) -> list[str]:
    n = len(s)
    if n <= WINDOW_SIZE:
        return []

    nums = [ENCODE[ch] for ch in s]

    bitmask = 0
    seen: set[int] = set()
    output: set[str] = set()

    for start in range(n - WINDOW_SIZE + 1):
        if start == 0:
            for index in range(WINDOW_SIZE):
                bitmask = (bitmask << 2) | nums[index]
        else:
            bitmask <<= 2
            bitmask |= nums[start + WINDOW_SIZE - 1]
            bitmask &= ~(3 << (2 * WINDOW_SIZE))

        if bitmask in seen:
            output.add(s[start:start + WINDOW_SIZE])
        else:
            seen.add(bitmask)

    return list(output)

File: 0187_repeated_dna_sequences/bitmask/go/solution.go
package repeateddnasequences

const windowSize = 10

func encode(ch byte) int {
	switch ch {
	case 'A':
		return 0
	case 'C':
		return 1
	case 'G':
		return 2
	default:
		return 3
	}
}

func findRepeatedDnaSequences(s string) []string {
	n := len(s)
	if n <= windowSize {
		return []string{}
	}

	nums := make([]int, n)
	for i := range s {
		nums[i] = encode(s[i])
	}

	bitmask := 0
	seen := make(map[int]struct{}, n)
	output := make(map[string]struct{})

	for start := 0; start <= n-windowSize; start++ {
		if start == 0 {
			for i := 0; i < windowSize; i++ {
				bitmask = (bitmask << 2) | nums[i]
			}
		} else {
			bitmask <<= 2
			bitmask |= nums[start+windowSize-1]
			bitmask &= ^(3 << (2 * windowSize))
		}

		if _, ok := seen[bitmask]; ok {
			output[s[start:start+windowSize]] = struct{}{}
		} else {
			seen[bitmask] = struct{}{}
		}
	}

	result := make([]string, 0, len(output))
	for fragment := range output {
		result = append(result, fragment)
	}
	return result
}
